### PR TITLE
Add JWT orchestrator

### PR DESCRIPTION
**Pull request by Avi-Eyal**
### What does this PR do?
- _What's changed? Why were these changes made?_
- _How should the reviewer approach this PR, especially if manual tests are required?_
- _Are there relevant screenshots you can add to the PR description?_

### What ticket does this PR close?
Resolves #[relevant GitHub issues, eg 76]

#### Change log
- [ ] The CHANGELOG has been updated, or
- [ ] This PR does not include user-facing changes and doesn't require a CHANGELOG update

#### Test coverage
- [ ] This PR includes new unit and integration tests to go with the code changes, or
- [ ] The changes in this PR do not require tests

#### Documentation
- [ ] Docs (e.g. `README`s) were updated in this PR, and/or there is a follow-on issue to update docs, or
- [ ] This PR does not require updating any documentation

#### API Changes
- [ ] The [OpenAPI spec](https://github.com/cyberark/conjur-openapi-spec) has been updated to meet new API changes (or an issue has been opened), or
- [ ] The changes in this PR do not affect the Conjur API

### CODE REVIEW

File: app/domain/authentication/authn_jwt/jwt_orchestrator.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+      end
+
+      private
+
+      def authenticate
+        "Jwt Orechestrator works!"
+      end
+    end
+  end
+end
```
Comment (tzheleznyak) [inline app/domain/authentication/authn_jwt/jwt_orchestrator.rb:21]: EOF

File: app/domain/authentication/authn_jwt/jwt_orchestrator.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
+require 'command_class'
+
+class JWTOrchestrator
```
Comment (tzheleznyak) [inline app/domain/authentication/authn_jwt/jwt_orchestrator.rb:3]: Add comment

Comment (nessiLahav) [reply app/domain/authentication/authn_jwt/jwt_orchestrator.rb:4]: Class definition is redundant here

File: app/domain/authentication/authn_jwt/jwt_orchestrator.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  module AuthnJwt
+
+    Authenticate ||= CommandClass.new(
```
Comment (nessiLahav) [inline app/domain/authentication/authn_jwt/jwt_orchestrator.rb:7]: Change Authenticate to JWTOrchestrator

File: app/domain/authentication/authn_jwt/jwt_orchestrator.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
+require 'command_class'
+
+class JWTOrchestrator
+  module AuthnJwt
```
Comment (nessiLahav) [inline app/domain/authentication/authn_jwt/jwt_orchestrator.rb:5]: Add Authentication module above

File: app/controllers/authenticate_controller.rb [hunk 5]
```diff
@@ -84,9 +84,21 @@ def authenticate(input = authenticator_input)
...
+    dummyResult = JWTOrchestrator::AuthnJwt::Authenticate.new.(
+      authenticator_input: authenticator_input
+    )
+    render status: 200, json: dummyResult
+  end
+
   # Update the input to have the username from the token and authenticate
   def authenticate_oidc
     params[:authenticator] = "authn-oidc"
+
```
Comment (nessiLahav) [inline app/controllers/authenticate_controller.rb:101]: delete

File: app/domain/authentication/authn_jwt/jwt_orchestrator.rb (hunk 3 above)
Comment (orenbm) [reply app/domain/authentication/authn_jwt/jwt_orchestrator.rb:7]: what does this class do? Does it orchestrate a JWT? Does it do it in one command? if so the name should be `OrchestrateJWT` to tell the action it does

Comment (orenbm) [reply app/domain/authentication/authn_jwt/jwt_orchestrator.rb:7]: the name of a command class should be self-explanatory. 
I must say that i don't understand what `OrchestrateJWT` is.
I would start by writing a description saying what this class does and how you should use it. The name should pop from that.

File: app/controllers/authenticate_controller.rb [hunk 6]
```diff
@@ -84,9 +84,21 @@ def authenticate(input = authenticator_input)
     handle_authentication_error(e)
   end
 
+  def authenticate_jwt
```
Comment (nessiLahav) [inline app/controllers/authenticate_controller.rb:87]: This function should be similar to `authenticate` function in the controller.

Something like:
```
 def authenticate_jwt(input = authenticator_input)
    authn_token = Authentication::JWTOrchestrator.new.(
      authenticator_input: input
    )
    content_type = :json
    if encoded_response?
      logger.debug(LogMessages::Authentication::EncodedJWTResponse.new)
      content_type = :plain
      authn_token = ::Base64.strict_encode64(authn_token.to_json)
      response.set_header("Content-Encoding", "base64")
    end
    render(content_type => authn_token)
  rescue => e
    handle_authentication_error(e)
  end
```

You will need to do some refactor to eliminate the duplicate code

File: app/domain/authentication/authn_jwt/jwt_orchestrator.rb (hunk 3 above)
Comment (Avi-Eyal) [reply app/domain/authentication/authn_jwt/jwt_orchestrator.rb:7]: Basically this class is supposed to orchestrate between different vendors configurations and authenticate according to each configuration 

Maybe we should call it OrchestrateJwtVendorConfiguration? @orenbm

Comment (orenbm) [reply app/domain/authentication/authn_jwt/jwt_orchestrator.rb:7]: does it create a Conjur access token at the end of its process?

Comment (Avi-Eyal) [reply app/domain/authentication/authn_jwt/jwt_orchestrator.rb:7]: As far as I know, it is

@nessiLahav can you confirm?

File: app/controllers/authenticate_controller.rb [hunk 7]
```diff
@@ -72,14 +72,17 @@ def authenticate(input = authenticator_input)
...
       enabled_authenticators: Authentication::InstalledAuthenticators.enabled_authenticators_str(ENV)
     )
-    content_type = :json
-    if encoded_response?
-      logger.debug(LogMessages::Authentication::EncodedJWTResponse.new)
-      content_type = :plain
-      authn_token = ::Base64.strict_encode64(authn_token.to_json)
-      response.set_header("Content-Encoding", "base64")
-    end
-    render(content_type => authn_token)
```
Comment (orenbm) [inline app/controllers/authenticate_controller.rb:82]: why did you remove this?

File: app/controllers/authenticate_controller.rb [hunk 8]
```diff
@@ -72,14 +72,17 @@ def authenticate(input = authenticator_input)
...
+    render_authn_token(authn_token)
+  rescue => e
+    handle_authentication_error(e)
+  end
+
+  def authenticate_jwt()
```
Comment (orenbm) [inline app/controllers/authenticate_controller.rb:80]:
```suggestion
  def authenticate_jwt
```

File: app/domain/authentication/authn_jwt/orchestrate_jwt_authentication.rb [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+    OrchestrateJwtAuthentication ||= CommandClass.new(
+    dependencies: {token_factory: TokenFactory.new},
+    inputs: %i[authenticator_input]
+  ) do
+    extend(Forwardable)
+    def_delegators(
+      :@authenticator_input, :account, :username
+    )
+      def call
```
Comment (orenbm) [inline app/domain/authentication/authn_jwt/orchestrate_jwt_authentication.rb:15]:
```suggestion
    OrchestrateJwtAuthentication ||= CommandClass.new(
      dependencies: {
        token_factory: TokenFactory.new
      },
      inputs: %i[authenticator_input]
    ) do
      extend(Forwardable)
      def_delegators(
        :@authenticator_input, :account, :username
      )
      
      def call
```

Comment (orenbm) [reply app/domain/authentication/authn_jwt/orchestrate_jwt_authentication.rb:15]: fixed indentation

**Comment by orenbm** [comment]
i'm ok with having this in a separate PR but note that you'll need to have an `Authentication::AuthnJwt::Authenticator` class that has a `valid?` method for the authenticator to be in the output of the `/authenticators` endpoint.

File: app/controllers/authenticate_controller.rb (hunk 7 above)
Comment (Avi-Eyal) [reply app/controllers/authenticate_controller.rb:82]: authenticate and authenticate_jwt use this same bock of code so i make a function (render_authn_token) out of it to avoid duplication of code

File: app/controllers/authenticate_controller.rb [hunk 10]
```diff
@@ -119,6 +122,17 @@ def authenticator_input
     )
   end
 
+  def render_authn_token(authn_token)
```
Comment (orenbm) [inline app/controllers/authenticate_controller.rb:125]:
```suggestion
  def rendered_authn_token(authn_token)
```

the method name should match the returned object

File: config/routes.rb [hunk 11]
```diff
@@ -33,6 +33,7 @@ def matches?(request)
         # i.e the request 'authn-oidc/:service_id/:account/authenticate' can be interpreted as
         # ':authenticator/:account/:id/authenticate'
         post '/authn-oidc(/:service_id)/:account/authenticate' => 'authenticate#authenticate_oidc'
+        post '/authn-jwt(/:service_id)/:account(/:identity)/authenticate' => 'authenticate#authenticate_jwt'
```
Comment (orenbm) [inline config/routes.rb:36]:
```suggestion
        post '/authn-jwt(/:service_id)/:account(/:id)/authenticate' => 'authenticate#authenticate_jwt'
```

change to `id` to match the request below

File: app/controllers/authenticate_controller.rb (hunk 10 above)
Comment (Avi-Eyal) [reply app/controllers/authenticate_controller.rb:125]: there is no returned object, the function renders the authn token, "render_authn_token" explains the action that will happen

Comment (orenbm) [reply app/controllers/authenticate_controller.rb:125]: pretty sure that `render` returns an object but i won't block the PR for this because it also makes sense to match the `render` call below.

**Review by orenbm — APPROVED** [review]
